test/unit/org/apache/cassandra/CassandraXMLJUnitResultFormatter.java

```diff
@@ -97,70 +104,107 @@ private static DocumentBuilder getDocumentBuilder() {
 
     /**
      * Element for the current test.
-     *
+     * <p>
      * The keying of this map is a bit of a hack: tests are keyed by caseName(className) since
      * the Test we get for Test-start isn't the same as the Test we get during test-assumption-fail,
      * so we can't easily match Test objects without manually iterating over all keys and checking
      * individual fields.
      */
-    private final Hashtable<String, Element> testElements = new Hashtable<String, Element>();
+    private final ConcurrentMap<String, Element> testElements = new ConcurrentHashMap<>();
 
     /**
      * tests that failed.
      */
-    private final Hashtable failedTests = new Hashtable();
+    private final ConcurrentMap<String, Element> failedTests = new ConcurrentHashMap<>();
 
     /**
      * Tests that were skipped.
      */
-    private final Hashtable<String, Test> skippedTests = new Hashtable<String, Test>();
+    private final ConcurrentMap<String, Test> skippedTests = new ConcurrentHashMap<>();
+
     /**
      * Tests that were ignored. See the note above about the key being a bit of a hack.
      */
-    private final Hashtable<String, Test> ignoredTests = new Hashtable<String, Test>();
+    private final ConcurrentMap<String, Test> ignoredTests = new ConcurrentHashMap<>();
+
     /**
      * Timing helper.
      */
-    private final Hashtable<String, Long> testStarts = new Hashtable<String, Long>();
+    private final ConcurrentMap<String, Long> testStarts = new ConcurrentHashMap<>();
+    private final ConcurrentMap<String, Long> testEnds = new ConcurrentHashMap<>();
+
+    private final ConcurrentSkipListMap<Long, Throwable> suiteEvents = new ConcurrentSkipListMap<>();
+    private final ConcurrentSkipListMap<Long, Throwable> testEvents = new ConcurrentSkipListMap<>();
+    private volatile ConcurrentSkipListMap<Long, Throwable> events = suiteEvents;
+
```

Comment:
nit: maybe helpful to have a comment describing what is represented by the key and value entries? e.g. for events the long is the event timestamp, and for testStarts/testEnds long is the start/end timestamp and the String is the test name. Just helps set context before one reads through the actual code.